### PR TITLE
create event classes for shopware events

### PR DESCRIPTION
**Pull request by uehler**
The Symfony Event System and therefore shopware 6 have those nice named events. For example when the shopware 6 product detail page was loaded, it will emit the event ProductPageLoadedEvent. It has getters for all the data it provides and your IDE helps you with code completion.
This pull request would create similar events for shopware 5.x, but I wanted to make sure this feature is wanted before I implement it. I also don't really know where to put those event classes in the shopware core, just tell me a good place.

You also won't have to copy the event names into your subscriber

```
use Enlight\Event\SubscriberInterface;

final class MySubscriber implements SubscriberInterface
{
    public static function getSubscribedEvents(): array
    {
        return [
            AddVoucherStartEvent::EVENT_NAME => 'mySubscriberCallback',
        ];
    }


    public function mySubscriberCallback(AddVoucherStartEvent $event)
    {
        echo $event->getCode();
    }
}
```

### 1. Why is this change necessary?
it is not, but might be helpful

### 2. What does this change do, exactly?
replaces all those generic Enlight_Event_EventArgs with named events, which provide getters and setters for the event data.

### 3. Describe each step to reproduce the issue or behaviour.
If you subscribe to any event in Shopware, you will get an Enlight_Event_EventArgs as parameter of your subscriber, but either you know which data is provided or you have to check where the event is emitted.
With the named events your IDE will provide you with code completion.

### 5. Which documentation changes (if any) need to be made because of this PR?
the event documentation should be extended, I would provide those changes too, if you like the idea of this PR

### 6. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have squashed any insignificant commits
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I have read the contribution requirements and fulfil them.

### CODE REVIEW

**Comment by shyim** [comment]
Maybe that could be a Task that can be done with https://github.com/rectorphp/rector
Just search for a Expr\_New and then create a class based on event name and pass the arguments there :sweat_smile:

**Comment by uehler** [comment]
I will have a look into it
where would be a good place for all those events?

**Comment by mitelg** [comment]
really nice idea! Also because it's not a break :blush: :+1: 
Maybe the events should be placed, where they are dispatched?

**Comment by uehler** [comment]
good idea
so for example I would place the AddVoucherStartEvent in engine/Shopware/Core/Events?

**Comment by mitelg** [comment]
yes, for example :+1: and also for each core class a subdirectory. but that is only my idea :blush: 
what do you think @shyim @soebbing ?

**Comment by uehler** [comment]
@shopwareBot yes, I will do it this week

File: engine/Shopware/Core/Events/Articles/GetArticleChartsEvent.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * trademark license. Therefore any rights, title and interest in
+ * our trademarks remain entirely with us.
+ */
+
+namespace Shopware\Core\Events\Articles;
+
+use Enlight_Event_EventArgs;
+use sArticles;
+
+class GetArticleChartsEvent extends Enlight_Event_EventArgs
```
Comment (mitelg) [inline engine/Shopware/Core/Events/Articles/GetArticleChartsEvent.php:30]: I'm wondering if we should use the correct translation "product" in the class name? :thinking: at some day devs might only use `Event::class` to subscribe to events, so we could get rid of another layer of false translations :wink:

**Comment by mitelg** [comment]
really nice work! :+1:

File: engine/Shopware/Core/Events/Articles/GetArticleChartsEvent.php (hunk 1 above)
Comment (uehler) [reply engine/Shopware/Core/Events/Articles/GetArticleChartsEvent.php:30]: good idea, I'll change that

Comment (mitelg) [reply engine/Shopware/Core/Events/Articles/GetArticleChartsEvent.php:30]: and also for the method names :wink:

File: engine/Shopware/Plugins/Default/Frontend/Seo/Bootstrap.php [hunk 2]
```diff
@@ -208,7 +208,8 @@ private function canBeMinified(Enlight_Controller_Action $controller): bool
         /** @var Enlight_Event_EventManager $events */
         $events = $this->get('events');
 
-        if ($event = $events->notifyUntil(__CLASS__ . '_canBeMinified', [
+        // TODO add event class and register namespace!
+        if ($event = $events->notifyUntil('Shopware_Plugins_Frontend_Seo_Bootstrap_canBeMinified', [
```
Comment (PaddyS) [inline engine/Shopware/Plugins/Default/Frontend/Seo/Bootstrap.php:212]: Hey @uehler,
will you fix this one too? :-)

Comment (uehler) [reply engine/Shopware/Plugins/Default/Frontend/Seo/Bootstrap.php:212]: i will look into this pr this week again
completely forgot about it 😅

Comment (shyim) [reply engine/Shopware/Plugins/Default/Frontend/Seo/Bootstrap.php:212]: Any new news?

**Pull request closed without merging**